lib/matplotlib/mlab.py

```diff
@@ -3404,44 +3404,46 @@ def griddata(x, y, z, xi, yi, interp='nn'):
 ##################################################
 # Linear interpolation algorithms
 ##################################################
-def less_simple_linear_interpolation(x, y, xi, extrap=False):
+_sentinel = object()
+def less_simple_linear_interpolation(x, y, xi, extrap=_sentinel):
     """
-    This function provides simple (but somewhat less so than
-    :func:`cbook.simple_linear_interpolation`) linear interpolation.
-    :func:`simple_linear_interpolation` will give a list of point
-    between a start and an end, while this does true linear
-    interpolation at an arbitrary set of points.
+    Linearly interpolate and extrapolate a function.
 
-    This is very inefficient linear interpolation meant to be used
-    only for a small number of points in relatively non-intensive use
-    cases.  For real linear interpolation, use scipy.
+    This function differs from `np.interp` by its ability to linearly
+    extrapolate beyond the lowest and highest x, and to handle multidimensional
+    output.
+
+    Parameters
+    ----------
+    x : a 1D array-like
+        The x-values where the interpolated function is known.
+    y : a 2D array-like
+        The values of the interpolated function.
+    xi : a 1D array-like
+        The x-values where to interpolate or extrapolate the function.
+
+    Returns
+    -------
+    The linearly interpolated and extrapolated values as a 2D array.
     """
     x = np.asarray(x)
     y = np.asarray(y)
-    xi = np.atleast_1d(xi)
-
-    s = list(y.shape)
-    s[0] = len(xi)
-    yi = np.tile(np.nan, s)
-
-    for ii, xx in enumerate(xi):
-        bb = x == xx
-        if np.any(bb):
-            jj, = np.nonzero(bb)
-            yi[ii] = y[jj[0]]
-        elif xx < x[0]:
-            if extrap:
-                yi[ii] = y[0]
-        elif xx > x[-1]:
-            if extrap:
-                yi[ii] = y[-1]
```

Comment:
The original is using the same extrapolation method as np.interp uses by default: setting the past-the-end values to the end values of the input array.  Are you changing the behavior because what mpl actually needs is linear extrapolation?

Reply:
Wow, I actually missed that point because my understanding of the contour code is that it indeed needs extrapolation (but I need to look at it again and run some examples).  If we don't actually need extrapolation then we can definitely get rid of this function in favor of using interp.

Reply:
I convinced myself that linear extrapolation was not needed because the call to `calc_label_rot_and_inline` is protected by `print_label` which is explicitly there to check that the contour line is long enough to host the label text, so we don't expect overflow anyways (so flat extrapolation is as good as anything).

Rewriting in terms of np.interp...